File: src/event_handler/event_handler_workflow.erl
-module(event_handler_workflow).

% elp:ignore W0012 W0040
-moduledoc false.

-export([
    execute/2
]).
-export([
    event_iterator/2
]).

-include_lib("temporal_sdk/include/workflow.hrl").

execute(_Context, _Input) ->
    start_execution(event_iterator, 1),
    start_activity(echo_activity, []),
    start_activity(echo_activity, [[], 5_000]),
    ok.

event_iterator(#{}, EventId) ->
    case await({event, {EventId, '_', '_', '_'}}) of
        {noevent, noevent} ->
            io:fwrite("NOEVENT~n", []);
        {ok, #{event_id := EventId, type := Type}} ->
            io:fwrite("~b: ~p~n", [EventId, Type]),
            event_iterator(#{}, EventId + 1)
    end.

File: lib/event_handler/event_handler_workflow.ex
defmodule EventHandler.Workflow do
  use TemporalSdk.Workflow
  @moduledoc false

  @impl true
  def execute(_context, _input) do
    start_execution(:event_iterator, 1)
    start_activity(:echo_activity, [])
    start_activity(:echo_activity, [[], 5_000])
    :ok
  end

  def event_iterator(%{}, event_id) do
    case await({:event, {event_id, :_, :_, :_}}) do
      {:noevent, :noevent} ->
        IO.puts("NOEVENT")

      {:ok, %{event_id: event_id, type: type}} ->
        IO.puts("#{event_id}: #{type}")
        event_iterator(%{}, event_id + 1)
    end
  end
end
